airrohr-firmware/mhz19.cpp

```diff
@@ -0,0 +1,150 @@
+    delay(200);
+    while(m_pSerialPort->available())
+      m_pSerialPort->read();
+
+    return true;
+  }
+  else
+  {
+    return false;
+  }
+}
+
+uint16_t CMHZ19Sensor::GetRange()
+{
+  return ppmMaxValue;
+}
+
+
+// returns 0xffff when reading fails
+uint16_t CMHZ19Sensor::ReadCO2Sensor(){
+  uint16_t ppm = 0xffff;
+  uint8_t response[packetLen]; 
+  static const uint8_t cmd[packetLen] = {0xFF,0x01,0x86,0x00,0x00,0x00,0x00,0x00,0x79}; // 2nd uint8_t is the sensor number
+  m_pSerialPort->write(cmd, sizeof(cmd));
+  if(!SynchronizeStream())
+  {
+#if DEBUG_OUTPUT    
+      Serial.print(F("Stream not synchronized "));
+#endif
+  }
+  else
+  {
+    uint8_t recvCnt = m_pSerialPort->readBytes((char*)response, sizeof(response));
+
+#if DEBUG_OUTPUT    
+    for(int8_t i = 0; i < min(recvCnt, packetLen); i++)
+    {
+      Serial.print(response[i], HEX);
+      Serial.write(' ');
+    }
+#endif
+  
+    if(recvCnt == sizeof(response)
+      && response[0] == 0xff
+      && response[1] == 0x86
+      && CalcCheckSum(response) == response[packetLen - 1]
+    )
+    {
+      uint16_t ppmRaw = (response[2] << 8) | response[3];
+      if(response[6] == 0x3A) // this marks the startup value (first ppm value will be 128 and then the max limit). After the startup phase this status goes to 0x35 and some time later to 0x34
+      {
+        // there will be values of 128, <maxValue>, 5, 301 and 400 on startup => ensure to ignore the non-relevant values!
+        // note: 397 was the smallest measured value, when using a range of 5000 ppm. This might me pure coincidence, but the values might also have a meaning
+        if(((ppmRaw % 1000) == 0) //&& (ppmRaw != 128) && (ppmRaw != 5) && (ppmRaw != 301) && (ppmRaw != 400)
+          && (ppmMaxValue != ppmRaw)
+         )
+        {
+          ppmMaxValue = ppmRaw;
+#if DEBUG_OUTPUT    
+          Serial.print(F("ppmMaxValue="));
+          Serial.println(ppmMaxValue);
+#endif
+        }      
+      }
+      else
+      {
+        ppm = ppmRaw;
+      }
+    }
+  }
+  return ppm;
+}
+
+uint8_t CMHZ19Sensor::CalcCheckSum(uint8_t *packet)
+{
+  uint8_t checksum = 0;
+  for(int8_t i = 1; i < (packetLen - 1); i++) //without the first uint8_t
+  {
+    checksum += packet[i];
+  }
+  return -checksum;
+}
+
+// synchronize with receive stream: first uint8_t must be 0xff
+// actually this consumes the rest of any dangling uint8_ts from the previous packet (will mainly happen on startup)
+bool CMHZ19Sensor::SynchronizeStream()
+{
+  const uint16_t millisecondsPerChar = 1.0/(mh_z19_baudrate / 10.0) * 1000.0 + 1.1; // 9600 Baud, 8N1
+  const uint16_t millisecondsPerPacket = packetLen * 1.0/(mh_z19_baudrate / 10.0) * 1000.0 + 10.1;// 9600 Baud, 8N1
+  delay(millisecondsPerPacket); // allow buffer to fill up
+  for(int8_t i = 0; i < (4*packetLen); i++)
+  {
+    int16_t c = m_pSerialPort->peek();
+    if(c == 0xff)
+    {
+      return true; // now we are 'probably' synchronized
```

Comment:
Do we want to check whether tje next byte is 0x86 ? Or whether the 8th byte matches the checksum ?

Reply:
No, that logic is done in readCO2Sensor(). synchronizeStream() only eats up any dangling bytes until the Start Of Frame is reached.